File: build-es-full.py
#!/usr/bin/env python3
"""Build es-full-map.json from ru-strings.json using Google Translate via deep-translator."""
import json
import re
import time
import sys
import os
from pathlib import Path
from deep_translator import GoogleTranslator

# Windows console UTF-8
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')
if hasattr(sys.stderr, 'reconfigure'):
    sys.stderr.reconfigure(encoding='utf-8', errors='replace')
os.environ.setdefault('PYTHONIOENCODING', 'utf-8')

ROOT = Path(__file__).parent
DELAY = 0.35
CHUNK = 4500

translator = GoogleTranslator(source='ru', target='es')

def is_good(ru, es):
    if not es or es == ru:
        return False
    if re.search(r'YOU USED ALL|MYMEMORY', es, re.I):
        return False
    if re.search(r'sección|clase de sección|clase div=', es, re.I):
        return False
    return True

def post_process(es: str) -> str:
    t = es
    t = re.sub(r'<clase de sección=', '<section class=', t, flags=re.I)
    t = re.sub(r'</sección>', '</section>', t, flags=re.I)
    t = re.sub(r'<clase div=', '<div class=', t, flags=re.I)
    t = re.sub(r'\boficina(s)?\b', lambda m: ('Cuenta' if m.group(0)[0]=='O' else 'cuenta') + ('s' if m.group(0).endswith('s') else ''), t, flags=re.I)
    t = re.sub(r'\bgabinete(s)?\b', lambda m: ('Cuenta' if m.group(0)[0]=='G' else 'cuenta') + ('s' if m.group(0).endswith('s') else ''), t, flags=re.I)
    t = re.sub(r'\bspread\b', 'gasto', t, flags=re.I)
    t = re.sub(r'\bdesenroscar\b', 'gastar', t, flags=re.I)
    t = re.sub(r'\bpor ruso\b', 'en español', t, flags=re.I)
    t = re.sub(r'«([^»]+)»', r'"\1"', t)
    t = re.sub(r'\bpor ruso\b', 'en español', t, flags=re.I)
    t = re.sub(r'\bGabinetes\b', 'Cuentas', t)
    t = re.sub(r'\bEstadísticas\b', 'Estadística', t)
    replacements = [
        (r'\bAdministrador de eventos\b', 'Events Manager'),
        (r'\badministrador de eventos\b', 'Events Manager'),
        (r'\bAdministrador de anuncios\b', 'Ads Manager'),
        (r'\badministrador de anuncios\b', 'Ads Manager'),
        (r'\badministrador(es)? comercial(es)?\b', 'Business Manager'),
        (r'\bgerente(s)? comercial(es)?\b', 'Business Manager'),
        (r'\bGabinetes\b', 'Cuentas'),
        (r'\bgabinetes\b', 'cuentas'),
        (r'\bprohibición\b', 'ban'),
        (r'\bestado: prohibición\b', 'status:ban'),
        (r'\bPUBLICIDAD\b', 'ADVERTISE'),
        (r'\bANUNCIAR\b', 'ADVERTISE'),
        (r'\bconjunto de anuncios\b', 'ad set'),
        (r'\bgrupo de anuncios\b', 'ad set'),
        (r'\bgrupos de anuncios\b', 'ad sets'),
        (r'\bregistraciones\b', 'registros'),
        (r'\bregistración\b', 'registro'),
        (r'\bdepa\b', 'depósito'),
        (r'\bdepas\b', 'depósitos'),
    ]
    for pat, rep in replacements:
        t = re.sub(pat, rep, t, flags=re.I)
    return t

def protect_templates(text):
    placeholders = []
    def repl(m):
        key = f'__PH{len(placeholders)}__'
        placeholders.append((key, m.group(0)))
        return key
    protected = re.sub(r'\$\{[^}]+\}', repl, text)
    return protected, placeholders

def restore_templates(text, placeholders):
    for key, val in placeholders:
        text = text.replace(key, val)
    return text

def translate_one(text, attempt=0):
    protected, ph = protect_templates(text)
    try:
        # deep-translator has length limits; split if needed
        if len(protected) <= CHUNK:
            raw = translator.translate(protected)
        else:
            parts = []
            for i in range(0, len(protected), CHUNK):
                parts.append(translator.translate(protected[i:i+CHUNK]))
                time.sleep(DELAY)
            raw = ''.join(parts)
        out = post_process(restore_templates(raw, ph))
        if not is_good(text, out):
            raise ValueError('bad output')
        return out
    except Exception as e:
        if attempt < 4:
            time.sleep(3 * (attempt + 1))
            return translate_one(text, attempt + 1)
        raise e

def load_seed():
    m = {}
    es_map = ROOT / 'es-map.json'
    if es_map.exists():
        for ru, es in json.loads(es_map.read_text(encoding='utf-8')).items():
            if is_good(ru, es):
                m[ru] = post_process(es)
    full = ROOT / 'es-full-map.json'
    if full.exists():
        for ru, es in json.loads(full.read_text(encoding='utf-8')).items():
            if is_good(ru, es):
                m[ru] = es
    return m

def apply_to_app(m):
    app_path = ROOT / 'app.js'
    content = app_path.read_text(encoding='utf-8')
    keys = sorted([k for k, v in m.items() if is_good(k, v)], key=len, reverse=True)
    applied = 0
    for k in keys:
        if k in content:
            content = content.replace(k, m[k])
            applied += 1
    content = content.replace('ru-RU', 'es-ES')
    content = re.sub(r'new Intl\.DisplayNames\(\["ru"\]', 'new Intl.DisplayNames(["es"]', content)
    content = re.sub(r'localeCompare\(([^,]+),\s*"ru"\)', r'localeCompare(\1, "es")', content)
    app_path.write_text(content, encoding='utf-8')
    cyr = len(re.findall(r'[а-яА-ЯёЁ]', content))
    return applied, cyr

def main():
    ru_list = json.loads((ROOT / 'ru-strings.json').read_text(encoding='utf-8'))
    ru_unique = list(dict.fromkeys(ru_list))
    m = load_seed()
    pending = [s for s in ru_unique if s not in m or not is_good(s, m.get(s, ''))]
    print(f'Seed: {len(m)}, pending: {len(pending)} / {len(ru_unique)}', flush=True)

    for i, s in enumerate(pending):
        if i % 20 == 0:
            preview = s[:55].replace('\n', ' ').encode('ascii', 'replace').decode('ascii')
            print(f'{i+1}/{len(pending)}: {preview}...', flush=True)
        try:
            m[s] = translate_one(s)
        except Exception as e:
            print(f'FAIL {s[:40]}: {e}', flush=True)
            m[s] = s
        if (i + 1) % 50 == 0:
            (ROOT / 'es-full-map.json').write_text(json.dumps(m, ensure_ascii=False, indent=2), encoding='utf-8')
            print('  checkpoint', flush=True)
        time.sleep(DELAY)

    for s in ru_unique:
        if s not in m:
            m[s] = s

    (ROOT / 'es-full-map.json').write_text(json.dumps(m, ensure_ascii=False, indent=2), encoding='utf-8')
    good = sum(1 for k, v in m.items() if is_good(k, v))
    print(f'Done: {len(m)} keys, {good} translated', flush=True)

    if '--apply' in sys.argv:
        applied, cyr = apply_to_app(m)
        print(f'Applied: {applied}, Cyrillic remaining: {cyr}', flush=True)

if __name__ == '__main__':
    main()

File: build-es-full.js
const fs = require('fs');
const https = require('https');

const DELAY_MS = 1500;
const RETRY_DELAY = 10000;

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function protectTemplates(text) {
  const placeholders = [];
  let i = 0;
  const protectedText = text.replace(/\$\{[^}]+\}/g, (m) => {
    const key = `__PH${i++}__`;
    placeholders.push({ key, value: m });
    return key;
  });
  return { protectedText, placeholders };
}

function restoreTemplates(text, placeholders) {
  let out = text;
  for (const { key, value } of placeholders) {
    out = out.split(key).join(value);
  }
  return out;
}

function postProcess(es) {
  let t = es;
  t = t.replace(/<clase de sección=/gi, '<section class=');
  t = t.replace(/<\/sección>/gi, '</section>');
  t = t.replace(/<clase div=/gi, '<div class=');
  t = t.replace(/\boficina(s)?\b/gi, (m) => (m[0] === 'O' ? 'Cuenta' : 'cuenta') + (m.endsWith('s') ? 's' : ''));
  t = t.replace(/\bgabinete(s)?\b/gi, (m) => (m[0] === 'G' ? 'Cuenta' : 'cuenta') + (m.endsWith('s') ? 's' : ''));
  t = t.replace(/\bspread\b/gi, 'gasto');
  t = t.replace(/\bdesenroscar\b/gi, 'gastar');
  t = t.replace(/\bpor ruso\b/gi, 'en español');
  t = t.replace(/«([^»]+)»/g, '"$1"');
  const replacements = [
    [/\bAdministrador de eventos\b/g, 'Events Manager'],
    [/\badministrador de eventos\b/g, 'Events Manager'],
    [/\bAdministrador de anuncios\b/g, 'Ads Manager'],
    [/\badministrador de anuncios\b/g, 'Ads Manager'],
    [/\badministrador(es)? comercial(es)?\b/gi, 'Business Manager'],
    [/\bgerente(s)? comercial(es)?\b/gi, 'Business Manager'],
    [/\bEstadísticas\b/g, 'Estadística'],
    [/\bGabinetes\b/g, 'Cuentas'],
    [/\bgabinetes\b/g, 'cuentas'],
    [/\bprohibición\b/gi, 'ban'],
    [/\bestado: prohibición\b/gi, 'status:ban'],
    [/\bPUBLICIDAD\b/g, 'ADVERTISE'],
    [/\bANUNCIAR\b/g, 'ADVERTISE'],
    [/\bconjunto de anuncios\b/gi, 'ad set'],
    [/\bgrupo de anuncios\b/gi, 'ad set'],
    [/\bgrupos de anuncios\b/gi, 'ad sets'],
    [/\bregistraciones\b/gi, 'registros'],
    [/\bregistración\b/gi, 'registro'],
    [/\bregi\b/gi, 'reg'],
    [/\bdepa\b/gi, 'depósito'],
    [/\bdepas\b/gi, 'depósitos'],
  ];
  for (const [re, rep] of replacements) t = t.replace(re, rep);
  if (t.includes('MYMEMORY WARNING')) {
    const parts = t.split('WARNING:');
    t = parts[parts.length - 1].trim();
  }
  return t;
}

function myMemory(text) {
  return new Promise((resolve, reject) => {
    const q = encodeURIComponent(text.slice(0, 4500));
    const url = `https://api.mymemory.translated.net/get?q=${q}&langpair=ru|es`;
    https.get(url, (res) => {
      let data = '';
      res.on('data', (c) => (data += c));
      res.on('end', () => {
        try {
          const j = JSON.parse(data);
          if (j.responseStatus === 200 || j.responseData?.translatedText) {
            resolve(j.responseData.translatedText);
          } else {
            reject(new Error(j.responseDetails || 'translate failed'));
          }
        } catch (e) {
          reject(e);
        }
      });
    }).on('error', reject);
  });
}

async function translateOne(text, attempt = 0) {
  const { protectedText, placeholders } = protectTemplates(text);
  try {
    const raw = await myMemory(protectedText);
    return postProcess(restoreTemplates(raw, placeholders));
  } catch (e) {
    if (attempt < 5) {
      await sleep(RETRY_DELAY * (attempt + 1));
      return translateOne(text, attempt + 1);
    }
    throw e;
  }
}

function applyToApp(map) {
  let content = fs.readFileSync('app.js', 'utf8');
  const sorted = Object.keys(map).filter((k) => map[k] && map[k] !== k).sort((a, b) => b.length - a.length);
  for (const k of sorted) content = content.split(k).join(map[k]);
  content = content.replace(/ru-RU/g, 'es-ES');
  content = content.replace(/new Intl\.DisplayNames\(\["ru"\]/g, 'new Intl.DisplayNames(["es"]');
  content = content.replace(/localeCompare\(([^,]+),\s*"ru"\)/g, 'localeCompare($1, "es")');
  fs.writeFileSync('app.js', content, 'utf8');
  return { applied: sorted.length, cyrillic: (content.match(/[а-яА-ЯёЁ]/g) || []).length };
}

async function main() {
  const ru = JSON.parse(fs.readFileSync('ru-strings.json', 'utf8'));
  const unique = [...new Set(ru)];
  let map = {};
  if (fs.existsSync('es-full-map.json')) {
    map = JSON.parse(fs.readFileSync('es-full-map.json', 'utf8'));
  } else if (fs.existsSync('es-map.json')) {
    map = JSON.parse(fs.readFileSync('es-map.json', 'utf8'));
  }

  const pending = unique.filter((s) => !map[s] || map[s] === s);
  console.log(`Pending: ${pending.length} / ${unique.length}`);

  for (let i = 0; i < pending.length; i++) {
    const s = pending[i];
    if (i % 10 === 0) console.log(`${i + 1}/${pending.length}: ${s.slice(0, 55)}...`);
    try {
      map[s] = await translateOne(s);
    } catch (e) {
      console.error('FAIL', s.slice(0, 50), e.message);
      map[s] = s;
    }
    if ((i + 1) % 25 === 0) fs.writeFileSync('es-full-map.json', JSON.stringify(map, null, 2), 'utf8');
    await sleep(DELAY_MS);
  }

  for (const s of unique) if (!map[s]) map[s] = s;
  fs.writeFileSync('es-full-map.json', JSON.stringify(map, null, 2), 'utf8');
  console.log('Wrote es-full-map.json with', Object.keys(map).length, 'entries');

  const { applied, cyrillic } = applyToApp(map);
  console.log('Applied', applied, 'translations');
  console.log('Cyrillic remaining:', cyrillic);
}

main().catch((e) => { console.error(e); process.exit(1); });
